### ISSUE TITLE
NA OFI: HG_Forward hangs when using Verbs provider with multiple contexts

### ISSUE THREAD

**Issue by vschhabra** [mercury-hpc/mercury#306]
We are running into a case of Mercury hanging when we use multiple contexts over verbs. With the help of debug tools in libfabric, we are seeing that mercury is hanging in its wait  on an fi_wait call and not waking up.  Th OFI logs show a few OFI send calls being issued and their corresponding completions getting reported but  thereafter progress stalls.

It seems from the OFI point of view,  mercury is not calling progress repeatedly. We can see it get called a few times only.

This same test works when we configure the sockets provider.

The test is invoking server to server RPC's. When the servers have multiple contexts, the hang is observed.

**Comment by a-ilango** [comment]
Some details on the issue from OFI perspective:

I put mercury in busy wait mode by setting the following and ran the test.
`hg_init_info.na_init_info.progress_mode = NA_NO_BLOCK;`

For the particular test, mercury uses OFI RxM/verbs in manual progress mode. Mercury opens 4 endpoints and 4 CQs. The CQs are bound to each of the endpoints (1:1). Mercury is supposed to progress all CQs but I see only one of them being progressed (at the end) and the test hangs.

Issue is reproducible with libfabric v1.8.0 and with master branch on Github.

**Comment by soumagne** [comment]
@carns is it related to the problems you've seen with verbs?

**Comment by carns** [comment]
It's possibly related.  We have a problem where performance collapses under certain threading configurations with libfabric v1.8.0 and verbs.  Just yesterday we collected some profiles comparing 1.7.1 (which performs fine) and 1.8.0 (which does not) for one the problematic configurations, and can see that the 1.8.0 build is spending the majority of it's CPU time in pthread_spin_lock().  I don't know if this is the same problem that @vschhabra is hitting, but in our case it definitely  looks like there is a poorly-placed spinlock somewhere in the code path now.  I'm hoping to make some headway figuring out exactly which spinlock is a problem soon.  I wasn't able to attribute the call path in our initial profiles.

**Comment by carns** [comment]
I opened a libfabric issue to see if someone has more insight there:

https://github.com/ofiwg/libfabric/issues/5244

The libfabric configure script could probably be hacked to disable spinlocks and/or userfaultfd to see if that helps.  I'm not likely to be able to look at it today myself, though.

**Comment by soumagne** [comment]
Thanks, very helpful. @a-ilango @vschhabra when running in busy wait mode, would you be able to send me a trace of where it hangs exactly? since in that mode there's no call to poll() or epoll() I would expect we could also pinpoint the mutex/spin lock where it stalls?

**Comment by a-ilango** [comment]
I think the issue pointed out by @carns is different from this one.

@soumagne, from what I've seen mercury tries to progress only one of the CQs when there are many that were opened. Because of this, endpoints which may receive a connection request are unable to handle the request. I think that's the reason for stall rather than a thread stuck in poll / spinlock.

In manual progress mode, all CQs that were bound to endpoints should be progressed. Is there a way to make mercury request auto-progress from OFI? I can test that to see if it works.

**Comment by carns** [comment]
The choice of manual/auto progress is hard coded right now, see:

https://github.com/mercury-hpc/mercury/blob/master/src/na/na_ofi.c#L156

You can toggle that value for the verbs provider from MANUAL to AUTO if you want to try it out.  The latter does work correctly on recent version of libfabric but I don't know what the performance implications are.

**Comment by soumagne** [comment]
@a-ilango I'm not sure I understand, you're the one responsible for making progress on the CQs since they are bound to each mercury context so you need to call HG_Progress on each of them. When using scalable endpoints (which is _not_ supported by verbs) you do have separate OFI queues per endpoint, one per HG context. So in the case of verbs you can only be in one of those 2 situations:
1. Multiple HG classes with multiple endpoints and one context, that you make progress on.
2. One single HG class and one endpoint, multiple HG contexts (sharing internally the same OFI queue) that you also make progress on.

Either way you have to make progress on the HG contexts that you create, Mercury won't do that automatically for you so it has to enter progress in that case, it's not possible to not progress a queue if you are calling HG_Progress() on that queue.

**Comment by shefty** [comment]
libfabric doesn't create any HG contexts, and isn't not aware of what middleware is sitting above it.  It's Mercury using libfabric, not the other way around.

The issue @a-ilango is referring to is that Mercury has allocated 4 libfabric CQs.  However, it is only reading from 1 of those.  Reading from the CQ is what drives progress at the libfabric level.

I would not assume that spinlock is necessarily the issue.  If a single thread is spinning on reading an empty CQ, the majority of its time would be spent acquiring/releasing a spinlock.  The lack of reading the other CQs could be causing the resulting performance drop.

**Comment by a-ilango** [comment]
@soumagne , do you mean CaRT (user of mercury IIUC) has to invoke the progress for the HG context?

> You can toggle that value for the verbs provider from MANUAL to AUTO if you want to try it out. The latter does work correctly on recent version of libfabric but I don't know what the performance implications are.

@carns I tried this but manual progress is still being requested via hints. Does using mercury in "busy wait" mode affect the progress mode being chosen?

Auto-progress mode would have lower performance than manual progress in situations where a system is fully subscribed (all cores in use).

**Comment by vschhabra** [comment]
Hi Jerome,

When you say that verbs does not support multiple endpoints, are you implying that CaRT has to have a separate implementation for Verbs? Currently it is the same as sockets and I think it is working over multiple end points but will have to verify.

Thanks
Vikram

**Comment by carns** [comment]
> @carns I tried this but manual progress is still being requested via hints. Does using mercury in "busy wait" mode affect the progress mode being chosen?

Oh, that's right.  Regardless of what provider you use, Mercury's na_ofi plugin will put it in manual progress mode if you use busy wait mode.

**Comment by a-ilango** [comment]
I removed busy wait mode (set `init_info.na_init_info.progress_mode = NA_DEFAULT`) and the test seems to be working now:
```
...
Guest_0@server-side checkin result - ret: 0, room_no: 1085.
main thread wait progress thread ...
```

So the test works with using RxM/verbs provider in auto-progress mode.

**Comment by soumagne** [comment]
@shefty right I never expected libfabric to create HG contexts :) we have the following mapping in mercury: an HG class is an OFI endpoint + an OFI CQ, HG contexts share the same CQ within an HG class or create their own CQ when using scalable endpoints but that's not the case here. `fi_cq_read()` gets called whenever we enter HG_Progress() on an HG context.
@a-ilango since you mentioned that you have 4 CQs and 4 endpoints, I assume that you have 4 HG classes with one HG context each in that case and you call `HG_Progress()` on each of those contexts. That means that you'd be calling `fi_cq_read()` on each of those CQs, I don't see anything that would prevent that from happening, assuming that you're making the correct `HG_Progress` calls. That's why I do not really understand why you say that only one CQ is progressed because if you have 4 calls to `HG_Progress()` on each of those contexts, each CQ has to be progressed (unless the other threads or something else is preventing the 3 other calls to `HG_Progress()` from being made)
@vschhabra no there's no need to have separate implementations, everything works the same way.

**Comment by soumagne** [comment]
@a-ilango yes when using busy mode it will set OFI's manual progress. Interesting that it works in auto progress, there's probably something in OFI's progress thread that does not get done the same way when doing manual progress...

**Comment by a-ilango** [comment]
>  I do not really understand why you say that only one CQ is progressed because if you have 4 calls to HG_Progress() on each of those contexts, each CQ has to be progressed (unless the other threads or something else is preventing the 3 other calls to HG_Progress() from being made)

In OFI, I see fi_cq_read being called only for one CQ. May be @vschhabra can answer what's happening on CaRT and whether it progresses all open HG contexts.

> there's probably something in OFI's progress thread that does not get done the same way when doing manual progress.

If auto-progress mode is requested, the RxM provider spawns a thread (per OFI endpoint) to drive the progress. So there's no need for upper layer software to issue progress calls. That's why it works. In manual progress mode it's the responsibility of the user (mercury or CaRT) to drive progress on all open OFI CQs.

**Comment by soumagne** [comment]
Well that's what tells me that something is wrong there because no matter what, Mercury needs to call `fi_cq_read()` and do something with what it gets from the OFI CQ... so whether or not progress gets made in the background should not matter since explicit progress needs to be done through the `HG_Progress()` call, i.e. the OFI CQ gets read and we push the corresponding mercury's operation to Mercury's CQ, that piece would not be done automatically, even with an OFI progress thread. So maybe there's some interdependency in that test and somehow it deadlocks if no progress is made? or something being done in OFI's progress thread is not done when doing manual progress?

**Comment by a-ilango** [comment]
The RxM/verbs auto-progress thread basically reads the verbs provider EQ so that any incoming connection events can be handled. In manual progress mode, the point at which the test got stuck was when a connection request was sent out but no response to that request was received (since no one was driving verbs EQ progress at the destination).

So when auto-progress is enabled, the one OFI CQ which was being read got completions as the connections got established automatically (RxM/verbs establishes connections to remote peers dynamically - on the first send message).

So my guess is other CQs are eventually read (by mercury invoked from CaRT) once the first CQ got completions?

**Comment by vschhabra** [comment]
I will add some instrumentation to the CaRT code to figure out what is going on.  Could be something in that space which was being hidden by the automatic progress mode.

Sent from my iPhone

On Sep 4, 2019, at 5:51 PM, Jerome Soumagne <notifications@github.com<mailto:notifications@github.com>> wrote:


Well that's what tells me that something is wrong there because no matter what, Mercury needs to call fi_cq_read() and do something with what it gets from the OFI CQ... so whether or not progress gets made in the background should not matter since explicit progress needs to be done through the HG_Progress() call, i.e. the OFI CQ gets read and we push the corresponding mercury's operation to Mercury's CQ, that piece would not be done automatically, even with an OFI progress thread. So maybe there's some interdependency in that test and somehow it deadlocks if no progress is made? or something being done in OFI's progress thread is not done when doing manual progress?

**Comment by soumagne** [comment]
OK this is definitely related I think. I would need to further look at how RxM handles connections but to me it seems weird to return FI_EAGAIN in that situation, i.e. the upper layer should not have to care about any connection being made under the hood since the purpose of the RxM layer is to ensure that? In other words, connection-related tasks that are not part of the semantics should not only be progressed when progress is called in my opinion... We could probably exit and retry the operation if FI_EAGAIN is returned but I don't think it would be great if we start having HG_Forward() calls that need to be retried... in some sense maybe that would be alright and that would give the opportunity for the user to progress other contexts? I would need to think more about it.

**Comment by shefty** [comment]
Apps already need to handle an operation returning EAGAIN.  The queues could be full, flow control could be backing the app off, etc.  Connection setup falls into this same area.  No data can be transferred until the connection completes.  Conceptually, it's the same as having 0 credits available to send.  Because OFI is intended to map to HW queues, we don't want to force the addition of a SW queue to avoid EAGAIN, since the apps typically have SW queues.

**Comment by soumagne** [comment]
OK yes that makes sense. I will look at fixing that at our level since we don't want to have to set OFI's progress to AUTO.

**Comment by johannlombardi** [comment]
Hi there,

Any update on this one?

**Comment by vschhabra** [comment]
Hi Johann

At our last Monday meeting this came up and Jerome said he would have the fix ready in two weeks.

Regard
Vikram

Sent from my iPhone

On Feb 8, 2020, at 3:00 AM, Johann Lombardi <notifications@github.com> wrote:

﻿

Hi there,

Any update on this one?

**Comment by soumagne** [comment]
Merged in 5d0cd776e140f76f27203d6f28d0fb3b092c9d03
